index/segment/fs/segment.go

```diff
@@ -0,0 +1,477 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fs
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"sync"
+	"unicode/utf8"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/generated/proto/fswriter"
+	"github.com/m3db/m3ninx/index"
+	"github.com/m3db/m3ninx/index/segment/fs/encoding"
+	"github.com/m3db/m3ninx/postings"
+	"github.com/m3db/m3ninx/postings/pilosa"
+	"github.com/m3db/m3ninx/postings/roaring"
+	"github.com/m3db/m3ninx/x"
+
+	"github.com/couchbase/vellum"
+	vregex "github.com/couchbaselabs/vellum/regexp"
+)
+
+var (
+	errReaderClosed            = errors.New("segment is closed")
+	errUnsupportedMajorVersion = errors.New("unsupported major version")
+	errNotImplemented          = errors.New("operation not implemented")
+
+	minByteKey = []byte{}
+	maxByteKey = []byte(string(utf8.MaxRune))
+)
+
+// NewSegmentOpts represent the collection of parameters to construct a Segment.
+type NewSegmentOpts struct {
+	MajorVersion  int
+	MinorVersion  int
+	Metadata      []byte
+	PostingsData  []byte
+	FSTTermsData  []byte
+	FSTFieldsData []byte
+
+	PostingsListPool postings.Pool
+}
+
+// NewSegment returns a new Segment backed by the provided options.
+func NewSegment(opts NewSegmentOpts) (Segment, error) {
+	if opts.MajorVersion != majorVersion {
+		return nil, errUnsupportedMajorVersion
+	}
+
+	metadata := fswriter.Metadata{}
+	if err := metadata.Unmarshal(opts.Metadata); err != nil {
+		return nil, err
+	}
+
+	if metadata.PostingsFormat != fswriter.PostingsFormat_PILOSAV1_POSTINGS_FORMAT {
+		return nil, fmt.Errorf("unsupported postings format: %v", metadata.PostingsFormat.String())
+	}
+
+	fieldsFST, err := vellum.Load(opts.FSTFieldsData)
+	if err != nil {
+		return nil, fmt.Errorf("unable to load fields fst: %v", err)
+	}
+	return &segment{
+		fieldsFST: fieldsFST,
+
+		opts:    opts,
+		numDocs: metadata.NumDocs,
+	}, nil
+}
+
+type segment struct {
+	sync.RWMutex
+	closed bool
+
+	fieldsFST *vellum.FST
```

Comment:
Would it be worth just using a sorted list for the fields instead of an FST since we don't need to search over the fields?

Reply:
No. A FST will compress much more than the sorted list, and will be faster than a sorted list at finding a field. Further we don't search over fields today, but no reason we can't support it in the future.

Reply:
I agree about compression, but I think in terms of performance it would be inferior. Can perhaps perform the benchmarks some time especially if we want to consider any B-Tree libraries.